src/main/java/io/confluent/connect/jdbc/source/TimestampIncrementingTableQuerier.java

```diff
@@ -229,6 +202,56 @@ public SourceRecord extractRecord() throws SQLException {
     return new SourceRecord(partition, offset.toMap(), topic, record.schema(), record);
   }
 
+  // Visible for testing
+  TimestampIncrementingOffset extractOffset(Schema schema, Struct record) {
+    final Timestamp extractedTimestamp;
+    if (timestampColumn != null) {
+      extractedTimestamp = (Timestamp) record.get(timestampColumn);
+      Timestamp timestampOffset = offset.getTimestampOffset();
+      assert timestampOffset != null && timestampOffset.compareTo(extractedTimestamp) <= 0;
+    } else {
+      extractedTimestamp = null;
+    }
+
+    final Long extractedId;
+    if (incrementingColumn != null) {
+      final Schema incrementingColumnSchema = schema.field(incrementingColumn).schema();
+      final Object incrementingColumnValue = record.get(incrementingColumn);
+      if (incrementingColumnValue == null) {
+        throw new ConnectException("Null value for incrementing column of type: " + incrementingColumnSchema.type());
+      } else if (isIntegralPrimitiveType(incrementingColumnValue)) {
+        extractedId = ((Number) incrementingColumnValue).longValue();
+      } else if (incrementingColumnSchema.name() != null && incrementingColumnSchema.name().equals(Decimal.LOGICAL_NAME)) {
+        final BigDecimal decimal = ((BigDecimal) incrementingColumnValue);
+        if (decimal.compareTo(LONG_MAX_VALUE_AS_BIGDEC) > 0) {
+          throw new ConnectException("Decimal value for incrementing column exceeded Long.MAX_VALUE");
+        }
+        if (decimal.scale() != 0) {
+          throw new ConnectException("Scale of Decimal value for incrementing column must be 0");
+        }
+        extractedId = decimal.longValue();
+      } else {
+        throw new ConnectException("Invalid type for incrementing column: " + incrementingColumnSchema.type());
+      }
+
+      // If we are only using an incrementing column, then this must be incrementing.
+      // If we are also using a timestamp, then we may see updates to older rows.
+      Long incrementingOffset = offset.getIncrementingOffset();
+      assert incrementingOffset == -1L || extractedId > incrementingOffset || timestampColumn != null;
+    } else {
+      extractedId = null;
+    }
+
+    return new TimestampIncrementingOffset(extractedTimestamp, extractedId);
+  }
+
+  private boolean isIntegralPrimitiveType(Object incrementingColumnValue) {
+    return incrementingColumnValue instanceof Long
```

Comment:
Probably doesn't matter much in practice, but any reason we switched to checking the value rather than the schema?